packages/endo/src/compartmap.js

```diff
@@ -0,0 +1,206 @@
+
+  const dependencies = [];
+  const children = [];
+  for (const name of keys(packageDescriptor.dependencies || {})) {
+    children.push(
+      // Mutual recursion ahead:
+      // eslint-disable-next-line no-use-before-define
+      gatherDependency(
+        readDescriptor,
+        graph,
+        dependencies,
+        packageLocation,
+        name,
+        tags
+      )
+    );
+  }
+
+  const { name, version } = packageDescriptor;
+  result.label = `${name}@${version}`;
+  result.dependencies = dependencies;
+  result.exports = inferExports(packageDescriptor, tags);
+
+  return Promise.all(children);
+};
+
+const gatherDependency = async (
+  readDescriptor,
+  graph,
+  dependencies,
+  packageLocation,
+  name,
+  tags
+) => {
+  const dependency = await findPackage(readDescriptor, packageLocation, name);
+  if (dependency === undefined) {
+    throw new Error(`Cannot find dependency ${name} for ${packageLocation}`);
+  }
+  dependencies.push(dependency.packageLocation);
+  await graphPackage(readDescriptor, graph, dependency, tags);
+};
+
+// This returns a graph whose keys are nominally URLs, one per package, with
+// values that are label: (an informative Compartment name, built as
+// ${name}@${version}), dependencies: (a list of URLs), and exports: (an object
+// whose keys are the thing being imported, and the values are the names of the
+// matching module, relative to the containing package's root, i.e. the URL
+// that was used as the key of graph). The URLs in dependencies will all exist
+// as other keys of graph.
+const graphPackages = async (
+  read,
+  packageLocation,
+  tags,
+  mainPackageDescriptor
+) => {
+  const memo = create(null);
+  const readDescriptor = packageLocation =>
+    readDescriptorWithMemo(memo, read, packageLocation);
+
+  if (mainPackageDescriptor !== undefined) {
+    memo[packageLocation] = Promise.resolve(mainPackageDescriptor);
+  }
+
+  const packageDescriptor = await readDescriptor(packageLocation);
+
+  tags = new Set(tags || []);
+  tags.add("import");
+  tags.add("endo");
+
+  if (packageDescriptor === undefined) {
+    throw new Error(
+      `Cannot find package.json for application at ${packageLocation}`
+    );
+  }
+  const graph = create(null);
+  await graphPackage(
+    readDescriptor,
+    graph,
+    {
+      packageLocation,
+      packageDescriptor
+    },
+    tags
+  );
+  return graph;
+};
+
+const translateGraph = (mainPackagePath, graph) => {
+  const compartments = {};
+
+  for (const [packageLocation, { label, dependencies }] of entries(graph)) {
+    const modules = {};
+    for (const packageLocation of dependencies) {
+      const { exports } = graph[packageLocation];
+      for (const [exportName, module] of entries(exports)) {
+        modules[exportName] = {
+          compartment: packageLocation,
+          module
+        };
+      }
+    }
+    compartments[packageLocation] = {
+      label,
+      root: packageLocation,
+      modules
```

Comment:
After studying the next PR in the series, I'd suggest adding comments here to address things I was confused about:

* point out that `modules` maps from the strings that this package can `import` to records that describe where that import is coming from; records which include the name of the compartment that hosts the imported module, and the relative name within that compartment of the module itself
* `modules` enumerates all the modules that this package is allowed to import, based on looking at all the packages its `package.json` declared a dependency upon, and enumerating all the modules those dependent packages say they can export. This surprised me (or rather, it makes sense in retrospect, but I keep forgetting this is the case) because I kind of expected that a table named "modules" and keyed to a given package would enumerate all the modules the package actually depends upon (which would require scanning *this* package, not other packages, and of course would be impossible to build in the face of dynamic imports with runtime-computed names).

Maybe a comment describing it as "available modules" or "importable modules" or "potential imports" might suffice.